sw/device/silicon_creator/lib/sigverify/sphincsplus/wots.c

```diff
@@ -0,0 +1,162 @@
+    kmac_shake256_absorb_words(ctx->pub_seed, kSpxNWords);
+    kmac_shake256_absorb_words(addr->addr, ARRAYSIZE(addr->addr));
+    kmac_shake256_absorb_words(out, kSpxNWords);
+    kmac_shake256_squeeze_start();
+    // This address change is located here for performance reasons; we update
+    // it while the Keccak core is processing.
+    spx_addr_hash_set(addr, i + 1);
+    HARDENED_RETURN_IF_ERROR(kmac_shake256_squeeze_end(out, kSpxNWords));
+  }
+
+  return kErrorOk;
+}
+
+/**
+ * Interprets an array of bytes as integers in base w.
+ *
+ * The NIST submission describes this operation in detail (section 2.5):
+ *   https://sphincs.org/data/sphincs+-r3.1-specification.pdf
+ *
+ * The caller is responsible for ensuring that `input` has at least
+ * `kSpxWotsLogW * out_len` bits available.
+ *
+ * This implementation assumes log2(w) is a divisor of 8 (1, 2, 4, or 8).
+ *
+ * @param input Input buffer.
+ * @param out_len Length of output buffer.
+ * @param[out] output Resulting array of integers.
+ */
+static_assert(8 % kSpxWotsLogW == 0, "log2(w) must be a divisor of 8.");
+static_assert(sizeof(uint8_t) <= kSpxWotsLogW,
+              "Base-w integers must fit in a `uint8_t`.");
+static void base_w(const uint8_t *input, const size_t out_len,
+                   uint8_t *output) {
+  size_t bits = 0;
+  size_t in_idx = 0;
+  uint8_t total;
+  for (size_t out_idx = 0; out_idx < out_len; out_idx++) {
+    if (bits == 0) {
+      total = input[in_idx];
+      in_idx++;
+      bits += 8;
+    }
+    bits -= kSpxWotsLogW;
+    output[out_idx] = (total >> bits) & (kSpxWotsW - 1);
+  }
+}
+
+/**
+ * Computes the WOTS+ checksum over a message (in base-w).
+ *
+ * The length of the checksum is `kSpxWotsLen2` integers in base-w; the caller
+ * must ensure that `csum_base_w` has at least this length.
+ *
+ * This implementation uses a 32-bit integer to store the checksum, which
+ * assumes that the maximum checksum value (len1 * (w - 1)) fits in that range.
+ *
+ * See section 3.1 of the NIST submission for explanation about the WOTS
+ * parameters here (e.g. `kSpxWotsLen2`):
+ *   https://sphincs.org/data/sphincs+-r3.1-specification.pdf
+ *
+ * @param msg_base_w Message in base-w.
+ * @param[out] csum_base_w Resulting checksum in base-w.
+ */
+static_assert(kSpxWotsLen1 * (kSpxWotsW - 1) <= UINT32_MAX,
+              "WOTS checksum may not fit in a 32-bit integer.");
+static void wots_checksum(const uint8_t *msg_base_w, uint8_t *csum_base_w) {
+  // Compute checksum.
+  uint32_t csum = 0;
+  for (size_t i = 0; i < kSpxWotsLen1; i++) {
+    csum += kSpxWotsW - 1 - msg_base_w[i];
+  }
+
+  // Make sure any expected empty zero bits are the least significant bits by
+  // shifting csum left.
+  size_t csum_nbits = kSpxWotsLen2 * kSpxWotsLogW;
+  csum <<= ((32 - (csum_nbits % 32)) % 32);
+
+  // Convert checksum to big-endian bytes and then to base-w.
+  csum = __builtin_bswap32(csum);
+  base_w((unsigned char *)&csum, kSpxWotsLen2, csum_base_w);
```

Comment:
nit `(uint8_t *)` for consistency.

Reply:
This cast uses the same reasoning as [here](https://github.com/lowRISC/opentitan/pull/17295#discussion_r1109521778).